src/Workspaces/Core/Portable/Workspace/Solution/ProjectId.cs

```diff
@@ -62,19 +64,19 @@ public override bool Equals(object obj)
             return this.Equals(obj as ProjectId);
         }
 
-        public bool Equals(ProjectId other)
+        public bool Equals(ProjectId? other)
         {
             return
                 !ReferenceEquals(other, null) &&
                 this.Id == other.Id;
         }
 
-        public static bool operator ==(ProjectId left, ProjectId right)
+        public static bool operator ==(ProjectId? left, ProjectId? right)
         {
-            return EqualityComparer<ProjectId>.Default.Equals(left, right);
+            return EqualityComparer<ProjectId>.Default.Equals(left!, right!);
```

Comment:
@333fred @jcouv @gafter I admit I didn't understand why I had to mark ! here, because EqualityComparer isn't annotated in our framework. I'm guessing the compiler is adding some implicit annotations here but it wasn't clear what they were. Is this documented somewhere?

Reply:
Because the `T` was substituted with `ProjectId`, which is a non-nullable reference type. And since you have an unannotated `EqualityComparer`, you don't get the `AllowNull` annotation on the `Equals` method.

Reply:
That confirms @agocke's guess at lunch -- I was mentally thinking that since we're on an unannotated framework in Roslyn we would always have fewer warnings than the annotated framework case. Had forgotten about things like this.

Reply:
:bulb: You should use `EqualityComparer<ProjectId?>` here instead of `EqualityComparer<ProjectId>`. It's a more accurate representation of intent and eliminates the need to use `!`.

Reply:
Why isn't it `EqualityComparer<ProjectId?>.Default` instead?